sdk/state/payment_test.go

```diff
@@ -854,3 +854,62 @@ func TestChannel_ProposeAndConfirmPayment_rejectIfChannelNotOpen(t *testing.T) {
 	_, err = senderChannel.ConfirmPayment(p)
 	require.EqualError(t, err, "validating payment: cannot confirm payment after an accepted coordinated close")
 }
+
+func TestChannel_enforceOnlyOneCloseAgreementAllowed(t *testing.T) {
+	localSigner := keypair.MustRandom()
+	remoteSigner := keypair.MustRandom()
+	localEscrowAccount := &EscrowAccount{
+		Address:        keypair.MustRandom().FromAddress(),
+		SequenceNumber: int64(101),
+		Balance:        int64(100),
+	}
+	remoteEscrowAccount := &EscrowAccount{
+		Address:        keypair.MustRandom().FromAddress(),
+		SequenceNumber: int64(202),
+		Balance:        int64(100),
+	}
+
+	senderChannel := NewChannel(Config{
+		NetworkPassphrase:   network.TestNetworkPassphrase,
+		Initiator:           true,
+		MaxOpenExpiry:       10 * time.Second,
+		LocalSigner:         localSigner,
+		RemoteSigner:        remoteSigner.FromAddress(),
+		LocalEscrowAccount:  localEscrowAccount,
+		RemoteEscrowAccount: remoteEscrowAccount,
+	})
+	receiverChannel := NewChannel(Config{
+		NetworkPassphrase:   network.TestNetworkPassphrase,
+		Initiator:           false,
+		MaxOpenExpiry:       10 * time.Second,
+		LocalSigner:         remoteSigner,
+		RemoteSigner:        localSigner.FromAddress(),
+		LocalEscrowAccount:  remoteEscrowAccount,
+		RemoteEscrowAccount: localEscrowAccount,
+	})
+
+	// Open channel.
+	m, err := senderChannel.ProposeOpen(OpenParams{
+		Asset:                      NativeAsset,
+		ExpiresAt:                  time.Now().Add(5 * time.Second),
+		ObservationPeriodTime:      10,
+		ObservationPeriodLedgerGap: 10,
+	})
+	require.NoError(t, err)
+	m, err = receiverChannel.ConfirmOpen(m)
+	require.NoError(t, err)
+	_, err = senderChannel.ConfirmOpen(m)
+	require.NoError(t, err)
+
+	// sender proposes payment
+	_, err = senderChannel.ProposePayment(10)
+	require.NoError(t, err)
+
+	// sender should not be able to propose a second payment until the first is finished
+	_, err = senderChannel.ProposePayment(20)
+	require.EqualError(t, err, "cannot start a new payment while an unfinished one exists")
+
+	// sender should not be able to propose coordinated close while unfinished payment exists
+	_, err = senderChannel.ProposeClose()
+	require.EqualError(t, err, "cannot propose coordinated close while an unfinished payment exists")
```

Comment:
In the PR description, you mentioned `A` should be able to do an uncoordinated close. Can you do it in this test just to show uncoordinated closes won't error?

Reply:
ah good question, it's actually already covered in the integration tests [(here)](https://github.com/stellar/experimental-payment-channels/blob/df1e4d83b13c7dea65836f31109db8740feeb202/sdk/state/integrationtests/state_test.go#L967-L993)

The uncoordinated close case isn't tested as a unit test because the sdk is currently supporting coordinated closes only by design (the happy case), so the integration tests are covering the uncoordinated closes cases (where something non-happy happens).

I added this check in the unit test though: https://github.com/stellar/experimental-payment-channels/pull/228/commits/2564ea5e34c0c7513f489957c478746c75b5848d
to make sure the original close agreement isn't lost (which means they can do an uncoordinated close with it if needed)